turnstile-example/turnstile/examples/cs478/current/stlc.rkt

```diff
@@ -385,3 +385,49 @@
   [(proj e i:nat) ≫
    -----------
    [≻ (tup-proj e i)]])
+
+(define-typerule (vals rec) ≫
+  [⊢ rec ≫ rec- ⇒ (~Rec [l = τ] ...)]
+  ------------------------------------
+  [⊢ (list- (cdr- (assoc- 'l rec-)) ...) ⇒ (× τ ...)])
+
+(define-type-constructor Sum #:arity = 2)
+
+(struct in-left (e) #:transparent)
+(struct in-right (e) #:transparent)
+
+(provide inl inr case Sum in-left in-right)
+
+(begin-for-syntax
+  (define (is-left tst-type sum-type)
+    (let ([τ ((current-type-eval) tst-type)]
+          [sum-τ ((current-type-eval) sum-type)])
+      (syntax-parse sum-τ
+        [(~Sum τ-left _) (type=? τ #'τ-left)])))
+
+  (define (is-right tst-type sum-type)
+    (let ([τ ((current-type-eval) tst-type)]
+          [sum-τ ((current-type-eval) sum-type)])
+      (syntax-parse sum-τ
+        [(~Sum _ τ-right) (type=? τ #'τ-right)]))))
+
+(define-typerule (inl e (~datum as) τ:type) ≫
```

Comment:
this is a pattern position, so I believe something like `(~and τ:type (~parse (~Sum τl _) #'τ.norm))` should work, instead of calling your helper fns.

Reply:
What does `norm` on this syntax class do?